compiler/src/dotty/tools/dotc/transform/init/Objects.scala

```diff
@@ -0,0 +1,1362 @@
+        eval(tree, thisV, klass)
+    }
+
+    thisV
+  }
+
+
+  /** Resolve C.this that appear in `klass`
+   *
+   * @param target  The class symbol for `C` for which `C.this` is to be resolved.
+   * @param thisV   The value for `D.this` where `D` is represented by the parameter `klass`.
+   * @param klass   The enclosing class where the type `C.this` is located.
+   * @param elideObjectAccess Whether object access should be omitted.
+   *
+   * Object access elission happens when the object access is used as a prefix
+   * in `new o.C` and `C` does not need an outer.
+   */
+  def resolveThis(target: ClassSymbol, thisV: Value, klass: ClassSymbol, elideObjectAccess: Boolean = false): Contextual[Value] = log("resolveThis target = " + target.show + ", this = " + thisV.show, printer, (_: Value).show) {
+    if target == klass then
+      thisV
+    else if target.is(Flags.Package) then
+      Bottom
+    else if target.isStaticObject then
+      val res = ObjectRef(target.moduleClass.asClass)
+      if target == klass || elideObjectAccess then res
+      else accessObject(target)
+    else
+      thisV match
+        case Bottom => Bottom
+        case Cold => Cold
+        case ref: Ref =>
+          val outerCls = klass.owner.lexicallyEnclosingClass.asClass
+          if !ref.hasOuter(klass) then
+            val error = "[Internal error] outer not yet initialized, target = " + target + ", klass = " + klass + Trace.show
+            report.error(error, Trace.position)
+            Bottom
+          else
+            resolveThis(target, ref.outerValue(klass), outerCls)
+        case RefSet(refs) =>
+          refs.map(ref => resolveThis(target, ref, klass)).join
+        case fun: Fun =>
+          report.error("[Internal error] unexpected thisV = " + thisV + ", target = " + target.show + ", klass = " + klass.show + Trace.show, Trace.position)
+          Bottom
+  }
+
+  /** Compute the outer value that correspond to `tref.prefix`
+   *
+   * @param tref    The type whose prefix is to be evaluated.
+   * @param thisV   The value for `C.this` where `C` is represented by the parameter `klass`.
+   * @param klass   The enclosing class where the type `tref` is located.
+   */
+  def outerValue(tref: TypeRef, thisV: Value, klass: ClassSymbol): Contextual[Value] =
+    val cls = tref.classSymbol.asClass
+    if tref.prefix == NoPrefix then
+      val enclosing = cls.owner.lexicallyEnclosingClass.asClass
+      resolveThis(enclosing, thisV, klass, elideObjectAccess = cls.isStatic)
+    else
+      if cls.isAllOf(Flags.JavaInterface) then Bottom
+      else evalType(tref.prefix, thisV, klass, elideObjectAccess = cls.isStatic)
+
+  def errorMutateOtherStaticObject(currentObj: ClassSymbol, otherObj: ClassSymbol)(using Trace, Context) =
+    val msg =
+      s"Mutating ${otherObj.show} during initialization of ${currentObj.show}.\n" +
+      "Mutating other static objects during the initialization of one static object is discouraged. " +
```

Comment:
Here you say "discouraged", in other messages earlier you say "forbidden".

Reply:
Which do you think is better?

Reply:
We now use `forbidden` for consistency, which also seems better than `discouraged`.